### PR TITLE
Refactor how we generate default json resolver

### PR DESCRIPTION
**Pull request by maffoo**
This came out of #3516; removing two levels of indentation makes the formatting change less drastic, but IMO also makes the code more understandable by removing the singleton `_ResolverCache` class.

### CODE REVIEW

File: cirq/protocols/json_serialization.py [hunk 1]
```diff
@@ -440,8 +428,9 @@ def _cirq_object_hook(d, resolvers: Sequence[JsonResolver]):
         raise ValueError("Could not resolve type '{}' "
                          "during deserialization".format(d['cirq_type']))
 
-    if hasattr(cls, '_from_json_dict_'):
-        return cls._from_json_dict_(**d)
+    from_json_dict = getattr(cls, '_from_json_dict_', None)
+    if from_json_dict is not None:
```
Comment (mpharrigan) [inline cirq/protocols/json_serialization.py:432]: why the change?

File: cirq/protocols/json_serialization.py [hunk 2]
```diff
@@ -36,183 +38,169 @@
...
-                'sympy.Pow': lambda args: sympy.Pow(*args),
-                'sympy.Float': lambda approx: sympy.Float(approx),
-                'sympy.Integer': sympy.Integer,
-                'sympy.Rational': sympy.Rational,
-                'complex': complex,
-            }
-        return self._crd
-
-
-RESOLVER_CACHE = _ResolverCache()
```
Comment (mpharrigan) [inline cirq/protocols/json_serialization.py:204]: technically a backwards-incompatible change but I think it should be fine because extensions (at least when I write them) use the `DEFAULT_RESOLVERS` global instead e.g. https://github.com/quantumlib/ReCirq/blob/master/recirq/serialization_utils.py#L225

File: cirq/protocols/json_serialization.py (hunk 1 above)
Comment (maffoo) [reply cirq/protocols/json_serialization.py:432]: The old code doesn't type check because mypy doesn't understand that `cls._from_json_dict_` is valid after the `hasattr` check.

File: cirq/protocols/json_serialization.py (hunk 2 above)
Comment (maffoo) [reply cirq/protocols/json_serialization.py:204]: We don't export `RESOLVER_CACHE` in the package-level `__init__.py` and it was only used in the associated tests so I wasn't considering it part of the public api. But I guess it's a bit ambiguous from the naming alone (no leading underscore).

Comment (balopat) [reply cirq/protocols/json_serialization.py:204]: I agree that this should be fine to remove, I added the BREAKING_CHANGE label so that we can list it in the release notes.

**Review by balopat — APPROVED** [review]
LGTM
